blog: OLake Fusion vs Spark Compaction (#404)
datazip-inc/olake-docs: blog/2026-04-22-spark-vs-fusion-compaction.mdx

---
slug: iceberg-compaction-spark-vs-fusion-benchmark

> title: "2x Faster Iceberg Compaction: How Fusion Beats Spark in CDC Workloads"

Reviewer: We need to change title. Cost is more attractive than speed in compaction.

OLake Fusion - Open-source 50% cheaper compaction than Spark.

description: "We benchmark Spark rewrite_data_files against OLake Fusion compaction on Apache Iceberg by running a full TPCH lineitem load from Postgres to GCP, applying 200k-record CDC batches every 2 minutes, and tracking TPC-H Query 6 performance, runtime, resource usage, and infrastructure cost."
tags: [iceberg, tpch, compaction, benchmark, spark, olake, fusion]
authors: [nayan]

> image: /img/blog/2026/4/fusion_vs_spark.webp

Reviewer: the image is not looking good. once ask Merlyn or design better image.

---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

![compaction diagram](/img/blog/2026/4/compaction-blog-img.webp)

Over the last few years, most data teams have either moved to a lakehouse architecture or are actively moving in that direction. That shift solves several legacy warehouse and data lake limitations, but migration alone is not the finish line. Once you're on a lakehouse, you still need to manage table health carefully to keep performance stable and costs under control.

> Every data lakehouse has a silent killer. It doesn't crash your pipeline. It doesn't throw errors. It just quietly makes your queries slower — day after day — until someone finally notices that what used to run in seconds now takes minutes. The culprit? Small files. And the fix? Compaction!

Reviewer 2: use less --

Author: updated the paragraphs and made sure more of the '-' is removed

> Most data lakehouse problems are loud. Pipelines fail and errors show up in logs. But there is one problem that stays completely silent. Your queries just get a little slower each day until one morning something that used to finish in seconds is now taking minutes. The reason is almost always small files, and the solution is compaction.

Reviewer: we need to tell why small files accumulate.

Author: mentioned how the small files were getting generated in the same section

> But here's the thing not all compaction is the same. The way you compact, how often you compact, and what engine you use to compact can make a dramatic difference in both query performance and operational cost. That's exactly what we set out to measure in this experiment.

Reviewer: ```suggestion
But here's the thing not all compaction is the same. The way you compact, how often you compact, and what engine you use to compact can make a dramatic difference in both query performance and resource+operational cost. That's exactly what we set out to measure in this experiment.
```

We ran a rigorous, benchmark comparing two compaction strategies on Apache Iceberg tables under continuous CDC ingestion — **Apache Spark's** `rewrite_data_files` and **OLake Fusion** Compaction — using identical data, identical resources, and identical ingestion (i.e. **OLake Go**). The results were eye-opening.

:::info Recommended background reading
If you want a conceptual foundation on compaction strategies and tiered maintenance for Iceberg, check out [OLake, Amoro, and the Iceberg Lakehouse Maintenance Playbook](/blog/olake-amoro-iceberg-lakehouse).
:::

In this blog, we skip the theory and go straight to the battlefield.

## Benchmarking Design: 

### 1. Data

> We used 300 GB of TPC-H data and focused exclusively on the lineitem table — the largest table in the TPC-H schema and the one that Query 6 operates on entirely. TPC-H Query 6 is a scan-heavy analytical query with filter predicates, making it extremely sensitive to file count and layout. It's the perfect canary for detecting compaction effectiveness.

Reviewer 2: only explain data not query use other section for it

Author: segregated the data and tpch in different sections

> We used 300 GB of TPC-H data and focused exclusively on the lineitem table — the largest table in the TPC-H schema. The table was **unpartitioned** in both runs, meaning all files were stored together without any partitioning, so any file count growth from CDC directly impacts how long queries take to scan the table.

Reviewer: but when i checked there were only 81gb went into full compaction right? so total data was 300 or 81?

Author: Total was 85 GB, but since we used 300TB TPCH data I had mentioned 300 and further mentioned only lineitem out of this was used. But have mentioned the size of lineitem table specifically as well.

### 2. Source

PostgreSQL served as the OLTP source, where we continuously issued row-level updates to emulate production-like CDC traffic.

> 

Reviewer: ```suggestion
PostgreSQL served as the OLTP source, where we continuously issued row-level updates every 2 mins to emulate production-like CDC traffic.

```

### 3. Destination

Apache Iceberg on Google Cloud Storage was the lakehouse destination, where frequent CDC commits gradually create small files and metadata overhead.

### 4. Catalog

Lakekeeper served as the Iceberg catalog, managing table metadata, snapshots, and commit coordination consistently across both environments.

### 5. Ingestion Pipeline

> For both environments, we used **OLake Go** as the ingestion engine — one of the fastest open-source tools for replicating data into Apache Iceberg table format. OLake handles both the initial full load and incremental CDC, and it's what makes the ingestion side of this experiment consistent and reproducible across both test environments.

Reviewer 2: use single line for olake

                    print(f"  Sleeping {round(sleep_seconds, 1)}s until next run ...")
                    time.sleep(sleep_seconds)
            next_run_monotonic += INTERVAL_SECONDS

            print(f"\n[{run_idx + 1}/{RUNS}] Running UPDATE on {SCHEMA}.{TABLE} - ~200k random rows ...")
            start = time.time()
            cursor.execute(update_sql)
            rows_affected = cursor.rowcount
            conn.commit()
            elapsed = round(time.time() - start, 2)
            print(f"[{run_idx + 1}/{RUNS}] Done. Rows updated: {rows_affected} | Time taken: {elapsed}s")

    except Exception as e:
        print(f"ERROR: {e}")
        if conn:
            conn.rollback()
            print("Transaction rolled back.")
        raise

    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()
        print("\nConnection closed.")
    ```

    </details>
- **OLake Go** was configured to sync every 2 minutes, picking up each batch of updates and writing them to the Iceberg destination as small incremental files.

This is as close to a real-world production CDC scenario as you can get in a benchmark.

### 6. Timeline

> The total experiment ran for **2 hours and 10 minutes**, structured deliberately:

Reviewer 2: afaik we run only 2 hours where 20 mints already made the gap that we discussed

Author: No we ran for 2 hrs 10 mins itself. First 10 mins were for ingestion and compaction started right at the point when 2 hours compaction started.

- **First 10 minutes** CDC starts and the first 5 ingestion cycles run (one every 2 minutes)

> - **After first 2 CDC ingestions** TPC-H Query 6 starts executing at regular intervals so we can track query time from the beginning.

Reviewer: can we mention the query?

- **After first 5 CDC ingestions** compaction kicks in to take on the growing small-file load.
- The **2-hour clock** starts the moment compaction kicks in — this is our primary observation window.

This structure gives us three distinct phases to observe: pre-compaction degradation, the immediate post-compaction response, and long-term steady-state performance.

> ## What Happens Without Compaction?

Reviewer: just make it 2-3 liner with link to that blog.

Before we compare Spark vs Fusion compaction head-to-head, it helps to see the baseline impact of not compacting at all. In our earlier benchmark, we ran TPC-H queries on Iceberg tables both without compaction and after compaction, and the query execution gap was massive. Read the full breakdown here: [Iceberg Compaction: How Much Faster Are TPC-H Queries?](/blog/iceberg-compaction-tpch-benchmark)

> The trend was unmistakable. Query execution time climbed steadily with each ingestion cycle. More files meant more metadata to read, more file handles to open, and more fragmented reads to execute. What started as a fast scan gradually became painfully slow — not because the data changed, but because the physical layout of the table had deteriorated.

Reviewer 2: more file handles to open -> more files to open

To keep this benchmark clean and unbiased, we provisioned **Fusion** and **Spark** with equivalent infrastructure and reused the **same TPCH query resources** in both runs so query conditions stayed identical.

Fusion shows one additional resource because that node runs the Fusion service itself, not extra compaction compute, so it does not create an unfair performance advantage.

<Tabs>
  <TabItem value="fusion-infra" label="OLake Fusion">

  | Node role | Instance type | vCPUs | RAM |
  |---|---|---|---|
  | Master | c4a-standard-8 | 8 | 32 GB |
  | Worker | c4a-highmem-16 | 16 | 128 GB |
  | Worker | c4a-highmem-16 | 16 | 128 GB |

  </TabItem>
  <TabItem value="spark-infra" label="Spark">

  | Node role | Instance type | vCPUs | RAM |
  |---|---|---|---|
  | Master | c4a-standard-8 | 8 | 32 GB |
  | Worker | c4a-highmem-16 | 16 | 128 GB |
  | Worker | c4a-highmem-16 | 16 | 128 GB |

  </TabItem>
  <TabItem value="tpch-infra" label="TPCH">

  | Node role | Instance type | vCPUs | RAM |
  |---|---|---|---|
  | Master | c4a-standard-8 | 8 | 32 GB |
  | Worker | c4a-standard-32 | 32 | 128 GB |
  | Worker | c4a-standard-32 | 32 | 128 GB |

</TabItem>
</Tabs>

> ## Compaction Setup and Configuration

Reviewer: lets remove tab view from here. I missed spark setup fully while reading as this section is quite long. 

tab setups are good to hide content which is either short or not-that-important-but-space-consuming-content.

Author: Removed the tab view

## Compaction Setup and Configuration

<Tabs>
  <TabItem value="fusion-compaction-config" label="OLake Fusion">

### OLake Fusion Configuration

OLake Fusion takes a tiered compaction approach, giving you three levels of optimization that can be scheduled independently based on how aggressively you want to maintain your tables.

All three compaction tiers are explained there in detail, so definitely check out the [OLake Iceberg Maintenance docs](/docs/iceberg-maintenance/optimization/overview).

For this benchmark, we used a **512 MB target file size**, so each tier behaves as follows:

- **Lite:** Files smaller than **64 MB** (1/8 of 512 MB) are merged into approximately **64 MB** output files. Fast, low-cost, and suitable for frequent runs during active ingestion.

> - **Medium:** Deletes are applied and files are merged into outputs that typically range from **64 MB to 512 MB**, depending on available data in each optimization cycle.

Reviewer 3: Please mention that they create positional delete files from equality delete files

Author: We have added a redirection link to our optimization page:
https://olake.io/docs/iceberg-maintenance/optimization/overview/
 itself which has all the information regarding all the types of compaction we offer. Again mentioning the same is required?

- **Full:** Performs a complete table rewrite to align output files close to the **512 MB** target, used when deeper cleanup and maximum query performance are required.

We used the following Fusion cluster configuration for running the compaction:

| Setting | Value |
|---|---|
| `spark-conf.spark.executor.instances` | 4 |
| `spark-conf.spark.executor.cores` | 7 |
| `spark-conf.spark.executor.memory` | 45g |
| `spark-conf.spark.executor.memoryOverhead` | 12g |
| `spark-conf.spark.driver.memory` | 18g |

> | `spark-conf.spark.driver.memoryOverhead` | 9g |

Reviewer 3: In fusion, mention dynamic allocation was false

Author: added the config

Fusion runs table optimization on Spark, and these `spark-conf` values size that workload: enough **executors and cores** to keep compaction moving in parallel, and enough **memory** overall so rewrite and merge steps are not constantly fighting for space.

### 1. How We Scheduled It

For this experiment, Fusion ran on the following schedule:

- Lite compaction: every 20 minutes
- Medium compaction: every 40 minutes

:::info Why Full compaction was skipped

> We intentionally skipped Full compaction in this benchmark because the observation window was short. In a small-period experiment like this, Full compaction would have had limited time to show its optimal long-term impact, so including it would likely not have changed the results meaningfully.

Reviewer: ```suggestion
* We intentionally skipped Full compaction in this benchmark because the observation window & total data size was quite small, where it would have had limited time to show its optimal long-term impact, so including it would likely not have changed the results meaningfully. 
* Also query performance was matched even without running Full. 
* Full is specially helpful when medium generates many files of the range target-file-size (TFS) to TFS/8 over period of 3-4 days or even a week.
```

  ```

</details>

  </TabItem>
</Tabs>

## TPC-H Query Execution

Alongside compaction, we ran **TPC-H Query 6 continuously** throughout the experiment with no wait time between runs - each new execution started immediately after the previous one finished. This allowed us to capture performance across all table states (before compaction, immediately after compaction, and later as new CDC files accumulated) and compute an unbiased average without overrepresenting any single phase.

We used the following Spark cluster configuration for TPC-H Query 6 runs:

| Setting | Value |
|---|---|
| `spark.executor.instances` | `12` |
| `spark.executor.cores` | `5` |
| `spark.executor.memory` | `15g` |
| `spark.executor.memoryOverhead` | `4g` |
| `spark.driver.memory` | `18g` |
| `spark.driver.memoryOverhead` | `10g` |

> | `spark.dynamicAllocation.enabled` | `false` |

Reviewer 3: are we sure for the TPCH query spark configuration dynamic allocation was false?

Author: yes I have cross checked that parameter in the spark submit for tpch query

  ```

</details>

## Benchmark Results: Who Actually Wins?

To make this comparison fair, we first made sure **query execution behavior stayed as aligned as possible** across both setups. Our goal was to avoid drawing conclusions from query-pattern differences and instead focus on what we actually wanted to compare: **compaction time, resource usage, and cost** for Spark vs Fusion.

Once query-time behavior was in a similar range, we had a stable baseline for evaluating the other constraints honestly. That let us compare both systems on operational efficiency, not just raw query outcomes.

### 1. TPC-H Query Time

> 

Reviewer: add table comparing spark and fusion average, min, max query time

In this section, we report the **minimum**, **maximum**, and **average** execution time of TPC-H Query 6 for both scenarios (Spark compaction and Fusion compaction). This gives a clear view of best-case, worst-case, and steady-state query behavior under continuous CDC and active maintenance.

<Tabs>

>   <TabItem value="fusion_tpch" label="Fusion TPCH Time">

Reviewer: this image doesn't have 60 and 120 min batch lines.

Reviewer: ```suggestion
  <TabItem value="fusion_tpch" label="Fusion TPCH Timing Details">
```

Author: Removed the batch line from other image so both the images are without the batch lines

  ![TPC-H Query 6 execution time — Fusion compaction](/img/blog/2026/4/fusion_tpch.webp)

  Over **134** consecutive runs with Fusion compaction enabled, TPC-H Query 6 was about **43s at best**, up to roughly **77s at worst**, with an **average near 56s**. Compaction keeps pulling the table back from small-file bloat, so query time does not stay stuck at the high end—it spikes and then settles again instead of drifting upward without bound.

  </TabItem>

>   <TabItem value="spark_tpch" label="Spark TPCH Time">

Reviewer: ```suggestion
  <TabItem value="spark_tpch" label="Spark TPCH Timing details">
```

  ![TPC-H Query 6 execution time — Spark compaction](/img/blog/2026/4/spark_tpch.webp)

  Over about **123** consecutive runs with Spark `rewrite_data_files` compaction on the same schedule, TPC-H Query 6 was about **45s at best**, up to roughly **83s at worst**, with an **average near 59s**. Periodic Spark compaction keeps the layout from degrading unchecked, so runtimes climb during small-file buildup then **drop back** instead of staying pinned near the worst case.

  </TabItem>
</Tabs>

### 2. Compaction Time

Here we compare how long each compaction approach took to run under the same ingestion and observation window. Use the tabs below for Fusion vs Spark.

<Tabs>
  <TabItem value="fusion-compaction-time" label="Fusion Compaction Time">

>   ![Fusion compaction time](/img/blog/2026/4/fusion_compaction.webp)

Reviewer 2: check image not correct, also validate spark query one as well

Author: Both the images checked they represent correct values itself.

Reviewer 2: The bar graph not look accurate verify once

>   <TabItem value="fusion-compaction-time" label="Fusion Compaction Time">

Reviewer: images are ok in tab-view but comparision has to be in the same table (Not in the tab view.)

**Final compaction-time verdict:** In this benchmark, **Fusion clearly outperforms Spark** - it finishes each compaction cycle in **4 mins 30 secs** on average vs **9 mins 18 secs** for Spark (**~2.07x faster**), and completes the full 6-run window in **27 mins 02 secs** vs **55 mins 47 secs**.

### 3. Cost

In this section, we compare the compaction cost for Spark and Fusion based on the resources used and the time each compaction workflow took to run. The objective is to show the operational trade-off clearly: not just which approach works, but which one delivers better efficiency per unit of compute spend.

Before calculating total compaction cost, here are the hourly rates for each instance type used in this benchmark:

| Instance type | Cost per hour (USD) |
| --- | --- |
| `e2-standard-8` | $0.32 |
| `c4a-standard-8` | $0.35 |
| `c4a-highmem-16` | $0.99 |
| `c4a-standard-32` | $1.60 |

Using the compaction infrastructure and total compaction durations from this benchmark:

| Engine | Resources included for compaction cost | Hourly infrastructure cost | Total compaction time | Total compaction cost |
| --- | --- | --- | --- | --- |

> | Fusion | `e2-standard-8`<br />`c4a-standard-8`<br />2 x `c4a-highmem-16` | $2.65/hour | 27 mins 02 secs | **$1.19** |

Reviewer 2: we are adding cost of fusion pod as well?, any reason?

Author: Cause Fusion pod is also consuming some resource, so this should also be involved in the cost comparison is what I think.

Author: Removing the cost for fusion pod as well just confirmed on it

| Spark | `c4a-standard-8`<br />2 x `c4a-highmem-16` | $2.33/hour | 55 mins 47 secs | **$2.17** |

#### Cost Summary

- **Fusion total compaction cost:** **$1.19**
- **Spark total compaction cost:** **$2.17**
- **Difference:** Spark costs about **$0.97 more** in this window

> - **Relative:** Fusion is about **44.9% lower cost** (or Spark is about **1.81x more expensive**) for compaction in this benchmark

Reviewer 3: 0.98 dollars, making it 45.16 % cheaper, check it once below towards the end as well

This is as close to a real-world production CDC scenario as you can get in a benchmark.

### 6. Timeline

The total experiment ran for **2 hours and 10 minutes**, structured deliberately:

- **First 10 minutes** CDC starts and the first 5 ingestion cycles run (one every 2 minutes)
- **After first 2 CDC ingestions** TPC-H Query 6 starts executing continuously so we can track query time from the beginning. Here is the query we used:

    ```sql
    SELECT
        SUM(l_extendedprice * l_discount) AS revenue
    FROM
        lineitem
    WHERE
        l_shipdate >= DATE '1994-01-01'
        AND l_shipdate <  DATE '1995-01-01'
        AND l_discount BETWEEN 0.05 AND 0.07
        AND l_quantity < 24
    ```

- **After first 5 CDC ingestions** compaction kicks in to take on the growing small-file load.
- The **2-hour clock** starts the moment compaction kicks in — this is our primary observation window.

> This structure gives us three distinct phases to observe: pre-compaction degradation, the immediate post-compaction response, and long-term steady-state performance.

Reviewer 2: till now we have not explained what actually benchmark is going to be, like we will keep query performance same, continue ingest data and compact making sure query performance is equal to other tools

Reviewer 2: also not talked about query going to be used to measure query perf

Author: Updated the information as a NOTE in that section

  ```

</details>

## TPC-H Query Execution

TPC-H Query 6 is a scan-heavy analytical query with filter predicates, making it extremely sensitive to file count and layout. It's the perfect canary for detecting compaction effectiveness.

Alongside compaction, we ran **TPC-H Query 6 continuously** throughout the experiment with no wait time between runs - each new execution started immediately after the previous one finished. This allowed us to capture performance across all table states (before compaction, immediately after compaction, and later as new CDC files accumulated) and compute an unbiased average without overrepresenting any single phase.

We used the following Spark cluster configuration for TPC-H Query 6 runs:

| Setting | Value |
|---|---|
| `spark.executor.instances` | 12 |
| `spark.executor.cores` | 5 |
| `spark.executor.memory` | 15g |
| `spark.executor.memoryOverhead` | 4g |
| `spark.driver.memory` | 18g |
| `spark.driver.memoryOverhead` | 10g |

> | `spark.dynamicAllocation.enabled` | false |

Reviewer 2: add a note that above configurations are to run query not compaction

Author: This is under TPCH Query Execution section so it indicates the configuration is for query execution

Reviewer 2: that is what am saying people read headlines not content

  ```

</details>

> ## Benchmark Results: Who Actually Wins?

Reviewer 2: highlight at end via note or something a basic table or single graph for both showing fusion winning

Author: we have already highlighted in each section that Fusion is performing better than Spark. In the conclusion as well its clearly mentioned. It will make it more repetitive if we keep repeating the same statement again.

Reviewer 2: people not read full blog they just see headlines

To make this comparison fair, we first made sure **query execution behavior stayed as aligned as possible** across both setups. Our goal was to avoid drawing conclusions from query-pattern differences and instead focus on what we actually wanted to compare: **compaction time, resource usage, and cost** for Spark vs Fusion.

Once query-time behavior was in a similar range, we had a stable baseline for evaluating the other constraints honestly. That let us compare both systems on operational efficiency, not just raw query outcomes.

### 1. TPC-H Query Time

In this section, we report the **minimum**, **maximum**, and **average** execution time of TPC-H Query 6 for both scenarios (Spark compaction and Fusion compaction). This gives a clear view of best-case, worst-case, and steady-state query behavior under continuous CDC and active maintenance.

<Tabs>
  <TabItem value="fusion_tpch" label="Fusion TPCH Timing Details">

  ![TPC-H Query 6 execution time — Fusion compaction](/img/blog/2026/4/fusion_tpch.webp)

>   {/* Over **134** consecutive runs with Fusion compaction enabled, TPC-H Query 6 was about **43s at best**, up to roughly **77s at worst**, with an **average near 56s**. Compaction keeps pulling the table back from small-file bloat, so query time does not stay stuck at the high end—it spikes and then settles again instead of drifting upward without bound.

Reviewer 2: while looking and both tpch graphs, (not checking the numbers): it shows there are more deep fluctuations in fusion query performance

Reviewer 2: can we make it little consistent like equal zoom for the both graphs

Author: The graph is plotted based on the actual values we got from the execution logs of the tpch queries.

Reviewer 2: Can we look at them once? Spark one shows consistency, while fusion one shows steep fluctuations.

  */}
  </TabItem>
</Tabs>

Comparison between Fusion and Spark TPC-H Query 6 execution time:

| Metric | Fusion | Spark |
| --- | --- | --- |
| Average query time | 56.00 s | 58.72 s |
| Min query time | 42.90 s | 45.62 s |
| Max query time | 76.71 s | 83.46 s |

### 2. Compaction Time

Here we compare how long each compaction approach took to run under the same ingestion and observation window. Use the tabs below for Fusion vs Spark.

<Tabs>
  <TabItem value="fusion-compaction-time" label="Fusion Compaction Time">

  ![Fusion compaction time](/img/blog/2026/4/fusion_compaction.webp)

  **Summary (Fusion):**
  - **Total compaction time:** 27 mins 02 secs
  - **Average per run:** 4 mins 30 secs
  - **Min:** 4 mins 11 secs
  - **Max:** 4 mins 46 secs

  </TabItem>
  <TabItem value="spark-compaction-time" label="Spark Compaction Time">

>   ![Spark compaction time](/img/blog/2026/4/spark_compaction.webp)

Reviewer 2: use same kind of graph for both

Author: In Line Chart it does not look good to mention what type of compaction we ran (like Lite, Medium) this only looks good in bar char. But For Spark since we used only one type of compaction the line chart looks good.

Reviewer 2: ask merlyn once, better person to check it
